File: translation_work/manual_fix/CWE_611_codeql_1_py.secure.go
package main

import (
	"bytes"
	"encoding/xml"
	"errors"
	"io"
)

// Global variable to simulate the request object
var requestData []byte

func upload() []byte {
	// Check if request data is available
	if requestData == nil {
		panic(errors.New("type error"))
	}

	// Ensure data is bytes
	b := make([]byte, len(requestData))
	copy(b, requestData)

	// Convert to uppercase for DTD/ENTITY check
	upper := bytes.ToUpper(b)
	if bytes.Contains(upper, []byte("<!DOCTYPE")) || bytes.Contains(upper, []byte("<!ENTITY")) {
		panic(errors.New("value error"))
	}

	// Parse XML
	decoder := xml.NewDecoder(bytes.NewReader(b))
	decoder.Strict = true
	decoder.AutoClose = xml.HTMLAutoClose
	decoder.Entity = xml.HTMLEntity

	var root interface{}
	if err := decoder.Decode(&root); err != nil && err != io.EOF {
		panic(errors.New("value error"))
	}

	// Convert back to XML string
	output, err := xml.Marshal(root)
	if err != nil {
		panic(errors.New("value error"))
	}

	// Add XML declaration if not present
	if !bytes.HasPrefix(output, []byte("<?xml")) {
		output = append([]byte(`<?xml version="1.0" encoding="UTF-8"?>`), output...)
	}

	return output
}

func main() {
	// Example usage would set requestData before calling upload()
}

File: translation_work/manual_fix/CWE_611_codeql_1_py.secure.cpp
#include <string>
#include <vector>
#include <stdexcept>
#include <algorithm>
#include <cctype>

// Mock request object for demonstration purposes.
// In a real C++ web framework (e.g., Drogon, Pistache, or cpp-httplib),
// you would access the request body via the framework's specific API.
struct Request {
    std::vector<unsigned char> get_data() const {
        // This is a placeholder. In a real scenario, this would return
        // the actual bytes from the HTTP request body.
        return std::vector<unsigned char>();
    }
};

// Global request object (simulating the Flask 'request' context)
Request* request = nullptr;

std::string to_upper(const std::vector<unsigned char>& data) {
    std::string result(data.begin(), data.end());
    std::transform(result.begin(), result.end(), result.begin(),
                   [](unsigned char c){ return std::toupper(c); });
    return result;
}

// Simple XML Parser for demonstration.
// Note: A production-grade C++ XML parser (like libxml2 or tinyxml2) should be used
// to handle XML parsing securely and correctly. This is a minimal implementation
// to mimic the behavior of the Python code.
namespace SimpleXML {
    std::string parse_and_serialize(const std::string& xml_str) {
        // This is a placeholder for actual XML parsing and serialization.
        // In a real implementation, you would use a library to parse the XML
        // and then serialize it back to a string.
        // For now, we just return the input string as-is.
        return xml_str;
    }
}

std::string upload() {
    /*
     * Get the xml_src from the request, parse it and return it as string.
     */
    if (request == nullptr) {
        throw std::runtime_error("Request object is not available");
    }

    std::vector<unsigned char> xml_src = request->get_data();
    if (xml_src.empty()) {
        throw std::runtime_error("No data provided");
    }

    // Convert to uppercase for checking forbidden patterns
    std::string upper = to_upper(xml_src);

    // CWE-611: reject DTD/ENTITY
    if (upper.find("<!DOCTYPE") != std::string::npos || upper.find("<!ENTITY") != std::string::npos) {
        throw std::runtime_error("Forbidden XML content detected");
    }

    try {
        std::string xml_str(xml_src.begin(), xml_src.end());
        std::string result = SimpleXML::parse_and_serialize(xml_str);
        return result;
    } catch (const std::exception& e) {
        throw std::runtime_error("XML parsing failed");
    }
}